quill-core/src/main/scala/io/getquill/context/TranslateContext.scala

```diff
@@ -0,0 +1,43 @@
+package io.getquill.context
+
+import io.getquill.NamingStrategy
+import io.getquill.idiom.Idiom
+import scala.annotation.tailrec
+import scala.language.experimental.macros
+
+trait TranslateContext {
+  this: Context[_ <: Idiom, _ <: NamingStrategy] =>
+
+  def translate[T](quoted: Quoted[T]): String = macro QueryMacro.translateQuery[T]
+  def translate[T](quoted: Quoted[Query[T]]): String = macro QueryMacro.translateQuery[T]
+  def translate(quoted: Quoted[Action[_]]): String = macro ActionMacro.translateQuery
+  def translate(quoted: Quoted[BatchAction[Action[_]]]): List[String] = macro ActionMacro.translateBatchQuery
+
+  def translateQuery[T](statement: String, prepare: Prepare = identityPrepare, extractor: Extractor[T] = identityExtractor): String = {
+    val params = prepareParams(statement, prepare)
+    if (params.nonEmpty) {
+      params.foldLeft(statement) {
+        case (expanded, param) => expanded.replaceFirst("\\?", param)
```

Comment:
Converting params from/to strings is not simple, I think we should keep the question marks and return a case class that has the string and the list of params (similarly to the mirror contexts)

Reply:
It's interesting that Slick people decided to log all printable query parameters in a completely separate logging statement (on a best effort basis) precisely for this reason.